File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 class BSTree{
    bool isReverse = true;
    private:
    stack<TreeNode*>st;
    void pushALL(TreeNode* root){
       while(root != NULL){
         if(isReverse == true){
            this->st.push(root);
            root = root->left;
         }
         else{
             this->st.push(root);
            root = root->right;
         }
       }
    }
    public:
    BSTree(TreeNode*root,bool reverse){
       this->isReverse = reverse;
       pushALL(root);
    }
    int next(){
        TreeNode*temp = st.top();
        st.pop();
        if(isReverse == true){
            if(temp->right != NULL){
                pushALL(temp->right);
            }
        }
        else{
            if(temp->left != NULL){
                pushALL(temp->left);
            } 
        }
        return temp->val;
    }

    bool hasNext(){
       if(!st.empty()){
        return true;
       }
       return false;
    }
 };
class Solution {
public:
 bool findTarget(TreeNode* root, int k) {
       BSTree l(root,true);  BSTree r(root,false);
       
       int left = l.next();
       int right = r.next();

       while(left<right){
         if(left+right == k){
            return true;
         }
         else if(left+right > k){
            right = r.next();
         }
         else{
            left = l.next();
         }
       }
       return false;
    }
};

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

     void preorder(TreeNode root,ArrayList<Integer> arr){
       if(root == null){
        return;
       }
       preorder(root.left,arr);
       arr.add(root.val);
       preorder(root.right, arr);
     }
    public boolean findTarget(TreeNode root, int k) {
        ArrayList<Integer>arr = new ArrayList<>();
        preorder(root, arr);
        int left = 0 , right = arr.size()-1;
         while(left<right){
            int sum = arr.get(left) + arr.get(right);
            if(sum == k) return true;
            if(sum < k){
                left++;
            }else{
                right--;
            }
         }

         return false;

    }
}